File: typescript/src/scenarios/s9-competing.ts
import { AckHandler, IMessageBus, IncomingMessage } from '../abstractions';
import {
  BusScenario,
  ScenarioResult,
  fail,
  nonce,
  pass,
  unsupported,
  waitUntil,
} from './scenario';

const MESSAGE_COUNT = 10;

/**
 * S9 — Competing consumers (work sharing, at-least-once). Two consumers attached
 * to the SAME queue share the load: each message is handled by exactly one of
 * them. This is the counterpoint to S6 fanout — same publish API, opposite
 * delivery semantics — and is the foundation of horizontal worker scaling.
 */
export const competingConsumers: BusScenario = {
  name: 'S9 competing consumers',
  description: `${MESSAGE_COUNT} messages are shared across 2 consumers, no duplicates.`,
  async run(bus: IMessageBus): Promise<ScenarioResult> {
    const t0 = performance.now();
    if (!bus.busCapabilities.supportsManualAck) {
      return unsupported(this.name, 'no manual ack on this broker', t0);
    }
    const topic = `mbc.s9.${nonce()}`;
    const queueId = `workers-${nonce()}`; // shared queue → competing consumers
    const byConsumer: string[][] = [[], []];

    const make = (i: number): AckHandler => async (m: IncomingMessage) => {
      byConsumer[i].push(m.body);
      await m.ack();
    };
    const subA = await bus.subscribe(topic, make(0), { subscriberId: queueId });
    const subB = await bus.subscribe(topic, make(1), { subscriberId: queueId });
    try {
      for (let i = 0; i < MESSAGE_COUNT; i++) {
        await bus.publish(topic, `job-${i}`);
      }
      const total = () => byConsumer[0].length + byConsumer[1].length;
      const ok = await waitUntil(() => total() >= MESSAGE_COUNT, 6000);

      const all = [...byConsumer[0], ...byConsumer[1]];
      const unique = new Set(all);
      if (!ok || all.length !== MESSAGE_COUNT) {
        return fail(this.name, `received ${all.length}/${MESSAGE_COUNT}`, t0);
      }
      if (unique.size !== MESSAGE_COUNT) {
        return fail(this.name, `duplicate delivery (${unique.size} unique)`, t0);
      }
      if (byConsumer[0].length === 0 || byConsumer[1].length === 0) {
        return fail(this.name, 'one consumer was starved (not balanced)', t0);
      }
      return pass(
        this.name,
        `split ${byConsumer[0].length}/${byConsumer[1].length}, no dupes`,
        t0,
      );
    } finally {
      await subA.unsubscribe();
      await subB.unsubscribe();
    }
  },
};

File: dotnet/src/Messaging.Scenarios/S9Competing.cs
using Messaging.Abstractions;
using static Messaging.Scenarios.ScenarioHelpers;

namespace Messaging.Scenarios;

/// <summary>
/// S9 — Competing consumers (work sharing, at-least-once). Two consumers attached
/// to the SAME queue share the load: each message is handled by exactly one of
/// them. This is the counterpoint to S6 fanout — same publish API, opposite
/// delivery semantics — and is the foundation of horizontal worker scaling.
/// </summary>
public sealed class CompetingConsumers : IBusScenario
{
    private const int MessageCount = 10;

    public string Name => "S9 competing consumers";
    public string Description => $"{MessageCount} messages are shared across 2 consumers, no duplicates.";

    public async Task<ScenarioResult> RunAsync(IMessageBus bus)
    {
        var t0 = StartClock();
        if (!bus.BusCapabilities.SupportsManualAck)
            return Unsupported(Name, "no manual ack on this broker", t0);

        var topic = $"mbc.s9.{Nonce()}";
        var queueId = $"workers-{Nonce()}"; // shared queue → competing consumers
        var byConsumer = new[] { new List<string>(), new List<string>() };
        var gate = new object();

        AckHandler Make(int i) => async m =>
        {
            lock (gate) byConsumer[i].Add(m.Body);
            await m.AckAsync();
        };

        var subA = await bus.SubscribeAsync(topic, Make(0), new SubscribeOptions { SubscriberId = queueId });
        var subB = await bus.SubscribeAsync(topic, Make(1), new SubscribeOptions { SubscriberId = queueId });
        try
        {
            for (var i = 0; i < MessageCount; i++)
                await bus.PublishAsync(topic, $"job-{i}");

            int Total() { lock (gate) return byConsumer[0].Count + byConsumer[1].Count; }
            var ok = await WaitUntilAsync(() => Total() >= MessageCount, TimeSpan.FromSeconds(6));

            List<string> all;
            int a, b;
            lock (gate)
            {
                all = byConsumer[0].Concat(byConsumer[1]).ToList();
                a = byConsumer[0].Count;
                b = byConsumer[1].Count;
            }
            var unique = new HashSet<string>(all);
            if (!ok || all.Count != MessageCount)
                return Fail(Name, $"received {all.Count}/{MessageCount}", t0);
            if (unique.Count != MessageCount)
                return Fail(Name, $"duplicate delivery ({unique.Count} unique)", t0);
            if (a == 0 || b == 0)
                return Fail(Name, "one consumer was starved (not balanced)", t0);

            return Pass(Name, $"split {a}/{b}, no dupes", t0);
        }
        finally
        {
            await subA.DisposeAsync();
            await subB.DisposeAsync();
        }
    }
}
